crates/ty_ide/src/completion.rs

```diff
@@ -14,44 +17,240 @@ pub struct Completion {
 pub fn completion(db: &dyn Db, file: File, offset: TextSize) -> Vec<Completion> {
     let parsed = parsed_module(db.upcast(), file);
 
-    let Some(target) = find_target(parsed, offset) else {
+    let Some(target) = CompletionTargetTokens::find(parsed, offset).ast(parsed) else {
         return vec![];
     };
 
     let model = ty_python_semantic::SemanticModel::new(db.upcast(), file);
-    let mut completions = model.completions(target.node());
-    completions.sort();
+    let mut completions = match target {
+        CompletionTargetAst::ObjectDot { expr } => model.attribute_completions(expr),
+        CompletionTargetAst::Scoped { node } => model.scoped_completions(node),
+    };
+    // Sort completions lexicographically, but make a leading `_` sort
+    // after any other leading byte.
```

Comment:
to add an additional wrinkle here: names that start and end with a double-underscore ("dunders") indicate attributes/methods that are given special treatment by the interpreter. It's less common to access these directly -- you would usually do `1 + 2` rather than `(1).__add__(2)`, for example -- but they're also not really "private" in the same way as other names that start with an underscore. So IMO the sorting should be:
1. Any names that do not start with an underscore
2. Dunder names
3. Any other names that start with an underscore

Reply:
Done!